fix: Segmentation and out-of-range errors.

xavieranc opened the pull request:
**`ScopeController.__init__`**: The number of channels is now stored in the `Parameter` object via the config and retrieved by `ScopeController` at this point. This avoids errors when the user uses the `.cfg` file to configure fewer than four channels (since `Configure` only creates two `Channels i` parameters; when the controller tries to access the third, for example during `update_fdr`, the parameter is not found).

**`ScopeController.execute_later`**: Implementation of a function that waits for the call stack to be emptied before executing an instruction. This function addresses segmentation issues when different threads attempt to modify parameters, especially when channels connect and request updates of the `'PV Status'` parameter.

**`ScopeController.muted`**: Addition of a function that allows executing an instruction after disconnection and before reconnection of the parameter-change signal (useful for changing a parameter without triggering the callback function).

**`ScopeController.set_channel_number`**: Implementation of a feature that allows changing the number of channels displayed on screen. This implementation addresses the following error: when the user specifies more fields on the command line than the number of channels requested in the `.cfg` file, an error is raised. The function therefore allows the user to modify the number of acquisition channels and is also called in the previously mentioned case to avoid the error.

Use of `ScopeController.muted` to control parameter changes originating from the code.

**`ScopeController.parameter_change`**: Removal of exception handling in the `if childName == "Acquisition.PV"` block since error handling is already implemented inside `ScopeController.update_fdr`.

**`ScopeControllerBase.update_status`**: Disconnection of the parameter-change signal before updating the statistics parameters, and reconnection afterward. On the first call of this function, the signal is normally not yet connected; the `try-except` structure allows handling the exception raised when Python tries to disconnect the signal while it is not yet connected.

**`MonitorStrategy.stop`**: Removal of an unnecessary line that was raising an error.

**`Configure.new_channel`**: The code generating the dictionary is moved into a separate function that can be called by `ScopeController`.

**`Configure.parse`**: Change in the order of function calls: `assemble_channel` must be called before `assemble_acquisition` so that the number of channels written in `'Acquisition'` is the effective numbers.

Commit:
fix: Segmentation and out-of-range errors.

echandler-anl commented:
Thanks for splitting out the changes.  What are the steps to reproduce the problems you were seeing that these changes address?

xavieranc commented:
Hi Elaine,

Regarding what I refer to as the "out-of-range" error, in your project on the main branch, you can simply edit the c2dv.cfg file and set COUNT=2 in the CHANNELS section. After making this change, launch the scope application. You’ll notice the following error appears:
KeyError: 'Parameter params has no child named Channel 3'

In my view, this happens because, in ScopeController, the line:
nchannels = kwargs.get("channels", 4)
sets nchannels to 4 regardless of the value specified in c2dv.cfg. For example, if you set COUNT=5, the fifth parameter is ignored in the scripts.

As for the segmentation faults, I couldn’t reproduce the errors directly from your code, since they were triggered by the code I wrote. However, some issues arose when the Channel connection callbacks attempted to modify parameter values (to display 'Connecting' or 'Connected' states), which led to recursive calls.

For now, I haven’t modified my first pull request, as it’s difficult to split the code. I suggest we address this current pull request first, and then I’ll update the first one based on the decisions we make.

echandler-anl commented on c2dataviewer/control/scope_controller_base.py:
```diff
@@ -313,6 +314,13 @@ def update_status(self):
         self.lastArrays = arraysReceived
         self.arrays = np.append(self.arrays, n)[-10:]
 
+        # Try to disconnect signal to prevent recursive calls of callback and then SegFaults.
```
Thanks for finding that c2dataviewer wasn't safely updating the parameter tree. It looks like pyqtgraph already has a builtin way to prevent recursive calls to the tree change callback.  Can you try using [treeChangeBlocker](https://pyqtgraph.readthedocs.io/en/latest/api_reference/parametertree/parameter.html#pyqtgraph.parametertree.Parameter.treeChangeBlocker) in these cases instead of disconnecting and reconnecting the callback.  They also have blockTreeChangeSignal/unblockTreeChangeSignal methods, but using treeChangeBlocker is preferable

xavieranc replied:
I did not use treeChangeBlocker because the doc says a signal is emitted at the end which is not what I expect. In fact from what I understood, using this fonction for one tree change does not change the behaviour. We can still try with block and unblockTreeChangeSignal.

echandler-anl commented on c2dataviewer/control/scopeconfig.py:
```diff
@@ -17,6 +18,10 @@
 """
 
+DEFAULT_MINIMUM_CHANNELS_NUMBER : int = 4
+DEFAULT_MINIMUM_WAVEFORMS_NUMBER : int = 4
```
The DEFAULT_*_WAVEFORMS_NUMBER variables appear to be unused, can you remove them?

Also, since the minimum and maximum waveform number doesn't change and "4" is the default number, not the maximum, consider having instead:

MIN_NCHANNELS = 1
MAX_NCHANNELS = 10
DEFAULT_NCHANNELS = 4

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -154,6 +158,52 @@ def __flatten_dict(dobj, kprefixs=[]):
+        '''
+        QtCore.QTimer.singleShot(0, instruction)
+
+    def muted(self, instruction : Callable[[], None]) -> None :
```
probably don't need this since pyqtgraph has a builtin way of blocking parameter tree change signals

echandler-anl commented on c2dataviewer/model/pvapy_plugins.py:
```diff
@@ -99,7 +101,7 @@ def poll(self):
         except pva.PvaException:
             #error because too many asyncGet calls at once.  Ignore
-            pass
+            logging.getLogger().exception('Failed to poll with PollStrategy.')
```
Changing this to a debug message, it isn't an error if it gets to this point

echandler-anl commented on c2dataviewer/model/pvapy_plugins.py:
```diff
@@ -166,11 +168,10 @@ def stop(self):
+            self.ctx.channel.setConnectionCallback(None)
         except PvaException:
-            pass
+            logging.getLogger().exception('Failed to stop Monitor.')
```
Change this to a debug message, it's not an error at this point.

echandler-anl commented on c2dataviewer/model/pvapy_plugins.py:
```diff
@@ -311,7 +312,7 @@ def async_get(self, data_callback, error_callback):
         except pva.PvaException:
             #error because too many asyncGet calls at once.  Ignore
-            pass
+            logging.getLogger().exception('Failed to get async with Channel.')
```
Changing this to a debug message

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -344,32 +392,25 @@ def parameter_change(self, params, changes):
+                        self.update_fdr()
+                    else:
+                        self.update_fdr(empty=True)
+                    self.reset_object()
```
This removes notifying the user of the error. What's the reason for the change?  If it doesn't fix an issue, then I'd like to revert it back to what it was.

xavieranc replied:
At first there was an issue with the following line in `MonitorStrategy.stop`:
  `self.ctx.channel.unsubscribe('monitorCallback')`
Since there was no subscription, the `stop` method was failing and the `update_fdr` also. The line above seems to be unnacessary, so I removed it.
I choose to remove this try-except structure because it did not give me any clue of this issue, while the try-except structures should. Moreover, I made sure that other try-excepts structures are placed at a more atomic level to control all of the new 395-401 lines. By doing so, the try-except of former lines 347&360 should not be triggered. For these reasons, I removed it. By testing the application, I did not have exceptions raised from the overall new block.

echandler-anl commented on c2dataviewer/control/scopeconfig.py:
```diff
@@ -134,7 +139,8 @@ def assemble_acquisition(self, section=None):
             {"name": "PV status", "type": "str", "value": "Disconnected", "readonly": True},
-            {"name": "Start", "type": "bool", "value": start}
+            {"name": "Start", "type": "bool", "value": start},
+            {'name' : 'Channels', 'type' : 'int', 'value' : self.counts, 'limits' : [1, 10]},
```
Can you move this parameter to the Config section, it doesn't fit exactly with the Acquisition section since it doesn't affect how it's acquiring data.  Also rename the parameter to "Channel count" to match configuration name.

use MIN and MAX constants here instead of hardcode 1 and 10

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -154,6 +158,52 @@ def __flatten_dict(dobj, kprefixs=[]):
             else:
                 yield sep.join(kprefixs + [k]), v
 
+    def execute_later(instruction : Callable[[], None]) -> None :
```
I don't see this used, please remove

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -154,6 +158,52 @@ def __flatten_dict(dobj, kprefixs=[]):
+        
+        :param number: The new number of sections that should be displayed.
+        '''
+        if number > 10 or number < 1 or type(number) is not int:
```
use MIN/MAX constants

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -154,6 +158,52 @@ def __flatten_dict(dobj, kprefixs=[]):
+        if previous_number > number:
+            self.channels = self.channels[: number]
+            while previous_number > number:
+                self.muted(lambda : self.parameters.child('Channel %s' % previous_number).remove())
```
instead of turning on/off signals for each call, maybe just turn them off for the entire function after checking the number range.

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -154,6 +158,52 @@ def __flatten_dict(dobj, kprefixs=[]):
+                previous_number -= 1
+        else:
+            while previous_number < number:
+                self.muted(lambda : self.parameters.insertChild(self.parameters.child('Statistics'), Configure.new_channel(previous_number + 1, self.color_pattern[previous_number], ['None'], 0)))
```
Can you add a comment in the code explaining what this part is doing, I think it is a little tricky since it depends on Statistics being right after the Channel sections.

echandler-anl commented on c2dataviewer/control/scopecontroller.py:
```diff
@@ -513,7 +558,7 @@ def stop_plotting(self):
+        self.muted(lambda : self.parameters.child("Acquisition").child("Start").setValue(0))
         # Stop data source
         self.model.stop()
 
```
Should add the new parameter for channel count the serialize function in ScopeController (this is further down in this file)

Commit:
fix: Hard-coded values, logging. feat: Channel count now in Config, serializing count.

echandler-anl committed:
revert: Restore try/except around PV update and inline buffer reset

Reverts the removal of error handling when updating the PV name, and
inlines the buffer reset logic instead of using reset_object().

echandler-anl committed:
refactor: Replace muted() with blockParamChangeSignal() using fn.SignalBlock
